core/src/main/java/org/apache/iceberg/io/AppenderBuilder.java

```diff
@@ -0,0 +1,136 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import org.apache.iceberg.MetricsConfig;
+import org.apache.iceberg.Schema;
+
+/**
+ * Interface which should be implemented by the data file format implementations. The {@link
+ * AppenderBuilder} will be parametrized based on the user provided configuration and finally the
+ * {@link AppenderBuilder#build(AppenderBuilder.WriteMode)} method is used to generate the appender
+ * for the specific writer use-cases. The following input should be handled by the appender in the
+ * specific modes:
+ *
+ * <ul>
+ *   <li>The appender's engine specific input type
+ *       <ul>
+ *         <li>{@link AppenderBuilder.WriteMode#APPENDER}
+ *         <li>{@link AppenderBuilder.WriteMode#DATA_WRITER}
+ *         <li>{@link AppenderBuilder.WriteMode#EQUALITY_DELETE_WRITER}
+ *       </ul>
+ *   <li>{@link org.apache.iceberg.deletes.PositionDelete} where the type of the row is the
+ *       appender's engine specific input type
+ *       <ul>
+ *         <li>{@link AppenderBuilder.WriteMode#POSITION_DELETE_WRITER}
+ *         <li>{@link AppenderBuilder.WriteMode#POSITION_DELETE_WITH_ROW_WRITER}
+ *       </ul>
+ * </ul>
+ *
+ * @param <B> type returned by builder API to allow chained calls
+ * @param <E> the engine specific schema of the input data
+ */
+public interface AppenderBuilder<B extends AppenderBuilder<B, E>, E> {
+  /** Set the file schema. */
+  B schema(Schema newSchema);
+
+  /**
+   * Set a writer configuration property.
+   *
+   * <p>Write configuration affects writer behavior. To add file metadata properties, use {@link
+   * #meta(String, String)}.
+   *
+   * @param property a writer config property name
+   * @param value config value
+   * @return this for method chaining
+   */
+  B set(String property, String value);
+
+  /**
+   * Set a file metadata property.
+   *
+   * <p>Metadata properties are written into file metadata. To alter a writer configuration
+   * property, use {@link #set(String, String)}.
+   *
+   * @param property a file metadata property name
+   * @param value config value
+   * @return this for method chaining
+   */
+  B meta(String property, String value);
+
+  /** Sets the metrics configuration used for collecting column metrics for the created file. */
+  B metricsConfig(MetricsConfig newMetricsConfig);
+
+  /** Overwrite the file if it already exists. The default value is <code>false</code>. */
+  B overwrite(boolean enabled);
```

Comment:
Do we need this to take a flag?  We should be consistent with the other flag based (e.g. `reuseContainers`).  We could either have the default be not-overwrite and expose `overwrite()` or make it a flag.

Reply:
Deprecated the `overwrite(boolean enabled)`.

The solution is a bit tricky. I need the old functionality until the deprecation allows for the removal of the old methods. In the end, I had to keep the `overwrite(boolean enabled)` in the new interface, but marked it for removal.